### PR TITLE
Add fallback cert support to Ingress v1

### PR DESCRIPTION
**Pull request by sunjayBhatia**
- when annotation projectcontour.io/enable-fallback-certificate == true,
fallback configured globally will be added
- since TLS secret is optional, if not provided then fallback used as
main secret and fallback
- if fallback configuration invalid, try to keep main TLS configuration
still configured
- errors surfaced via logs since we don't have status

Fixes: #2453
Fixes: #4556

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
WIP, wanted to get it out for review early, still need to test things out more and write an e2e test

main odd thing to review is to look at what happens when no TLS secret is provided in Ingress TLS section (since it is technically optional) and fallback is enabled, right now I have it so that the fallback cert is added as the main vhost TLS secret *and* the fallback (for requests w/o SNI)

File: internal/dag/ingress_processor.go [hunk 1]
```diff
@@ -77,34 +82,102 @@ func (p *IngressProcessor) Run(dag *DAG, source *KubernetesCache) {
...
+				fallbackCertEnabled = false
+			}
+		}
+
+		for _, tls := range ing.Spec.TLS {
+			var secret *Secret
+			if tls.SecretName == "" {
+				if fallbackCertEnabled {
+					// If no provided secret, use fallback cert as main vhost secret.
+					secret = fallbackCert
```
Comment (skriss) [inline internal/dag/ingress_processor.go:142]: Is this the same behavior as for HTTPProxy? (started looking but if you've already tracked this down I'll spare myself the spelunk).

**Comment by skriss** [comment]
> main odd thing to review is to look at what happens when no TLS secret is provided in Ingress TLS section (since it is technically optional) and fallback is enabled, right now I have it so that the fallback cert is added as the main vhost TLS secret and the fallback (for requests w/o SNI)

I haven't spent a lot of time with this code, so I might be missing something, but it seems to me like you could either (a) do what you've done here; or (b) not add a filterchain for the specific vhost, and you'd effectively end up with the same user experience (since in b, requests for that vhost would match the fallback filterchain which would have the same cert and still have all the routes). Does that sound right to you? If so, then I think what you've got here makes sense and is probably a bit more efficient than (b).

File: internal/dag/ingress_processor.go (hunk 1 above)
Comment (sunjayBhatia) [reply internal/dag/ingress_processor.go:142]: HTTPProxy requires you to specify a TLS cert so you always get a "main" TLS cert for most requests and the fallback for ones w/o SNI

with Ingress you don't have to provide a cert so this solution is kind of stretching of the meaning of "fallback" to extend to that use case

**Comment by sunjayBhatia** [comment]
> > main odd thing to review is to look at what happens when no TLS secret is provided in Ingress TLS section (since it is technically optional) and fallback is enabled, right now I have it so that the fallback cert is added as the main vhost TLS secret and the fallback (for requests w/o SNI)
> 
> I haven't spent a lot of time with this code, so I might be missing something, but it seems to me like you could either (a) do what you've done here; or (b) not add a filterchain for the specific vhost, and you'd effectively end up with the same user experience (since in b, requests for that vhost would match the fallback filterchain which would have the same cert and still have all the routes). Does that sound right to you? If so, then I think what you've got here makes sense and is probably a bit more efficient than (b).

yeah, you would get the same ux with (a) and (b) for a user making requests, with just a minor hit that you end up going through the catchall "any TLS" filterchain+fallback route

originally I sortof tried to do (b) but we would need a few small/fiddly logic changes to the dag and xdscache to make it ok to have a secure vhost that does not do passthrough TLS or have a secret

**Comment by tsaarni** [comment]
In the [docs]( https://projectcontour.io/docs/main/config/tls-termination/#fallback-certificate) the fallback certificate is currently explained through the use case where client does not set SNI. Using the fallback cert also when SNI is set would be small exception to that, a bit of a stretch like said previously, but of course it is just one definition.

Technically user could configure `secretName` with a reference to the fallback certificate secret (assuming they know the name of it), instead of configuring `projectcontour.io/enable-fallback-certificate: true` annotation, like they need to do for `HTTPProxy` (as stated previously). But if I understood if correctly, the motivation with the automatic behavior is to avoid an error case since the field is optional?  One can still have similar errors if (1) leaving `secretName` field empty and not setting `projectcontour.io/enable-fallback-certificate: true` or if (2) fallback certificate secret name has not been configured for the deployment or delegation rule is not there.

The description for leaving field `Ingress.spec.tls.secretName` empty says _"Field is left optional to allow TLS routing based on SNI hostname alone"_ so I wonder if that option is really meant for TLS passthrough?

That said, I do not have any objections for the suggested implementation! :)  It seems to match what [NGINX](https://kubernetes.github.io/ingress-nginx/user-guide/tls/#default-ssl-certificate) does, if I understood the description correctly (did not test).

**Comment by buroa** [comment]
Ping @sunjayBhatia, can we get this merged?

**Pull request closed without merging**